python/ray/serve/config.py

```diff
@@ -138,6 +149,252 @@ def _to_deployment_mode(cls, v: Union["ProxyLocation", str]) -> DeploymentMode:
             raise ValueError(f"Unrecognized `ProxyLocation`: {v}.")
 
 
+@PublicAPI(stability="stable")
+class BaseRayActorOptionsModel(BaseModel):
+    """Options with which to start a replica actor."""
+
+    runtime_env: dict = Field(
+        default={},
+        description=(
+            "This deployment's runtime_env. working_dir and "
+            "py_modules may contain only remote URIs."
+        ),
+    )
+    num_cpus: float = Field(
+        default=None,
+        description=(
+            "The number of CPUs required by the deployment's "
+            "application per replica. This is the same as a ray "
+            "actor's num_cpus. Uses a default if null."
+        ),
+        ge=0,
+    )
+    num_gpus: float = Field(
+        default=None,
+        description=(
+            "The number of GPUs required by the deployment's "
+            "application per replica. This is the same as a ray "
+            "actor's num_gpus. Uses a default if null."
+        ),
+        ge=0,
+    )
+    memory: float = Field(
+        default=None,
+        description=(
+            "Restrict the heap memory usage of each replica. Uses a default if null."
+        ),
+        ge=0,
+    )
+    object_store_memory: float = Field(
+        default=None,
+        description=(
+            "Restrict the object store memory used per replica when "
+            "creating objects. Uses a default if null."
+        ),
+        ge=0,
+    )
+    resources: Dict = Field(
+        default={},
+        description=("The custom resources required by each replica."),
+    )
+    accelerator_type: str = Field(
+        default=None,
+        description=(
+            "Forces replicas to run on nodes with the specified accelerator type."
+        ),
+    )
+
+    @validator("runtime_env")
+    def runtime_env_contains_remote_uris(cls, v):
+        # Ensure that all uris in py_modules and working_dir are remote
+
+        if v is None:
+            return
+
+        uris = v.get("py_modules", [])
+        if "working_dir" in v:
+            uris.append(v["working_dir"])
+
+        for uri in uris:
+            if uri is not None:
+                try:
+                    parse_uri(uri)
+                except ValueError as e:
+                    raise ValueError(
+                        "runtime_envs in the Serve config support only "
+                        "remote URIs in working_dir and py_modules. Got "
+                        f"error when parsing URI: {e}"
+                    )
+
+        return v
+
+
+NumReplicasAnnotatedType = Annotated[
```

Comment:
I'm explicitly defining the type annotations for each field because I want to override the default values in `ApplyServeDeploymentModel(BaseDeploymentModel)` without having to redefine the metadata. Currently there's no good support in Pydantic for this.

Reply:
oof, this looks somewhat annoying to maintain. I don't know of a better solution though.

Reply:
Why do we need to override the defaults? That has a funny smell to it

Reply:
+1, I'm not sure why we need to override the defaults.

An alternative to this approach is to [construct the schemas dynamically](https://docs.pydantic.dev/latest/concepts/models/#dynamic-model-creation). You could create a factory function that constructs schema with different defaults.

Reply:
I've found an alternative to overriding the defaults. The goal is to know which fields the user explicitly set, but pydantic models have the built-in functionality of tracking this (either through `.dict(exclude_unset)` or `.__fields_set__`). I've changed the codebase to take advantage of pydantic's feature, so now there's no longer a need to override the defaults of the user-facing model.